Handle history archive magnetlink messages

0x-r4bbit opened the pull request:
This introduces the ability for status notes to handle community
history archive magnetlinks. To make this work, a few things are needed:

1. A new database table has been introduced to store message archive
   hashes. This is necessary so status nodes can determine whether or
   not they need to download a certain archive
2. The messenger's `handleRetrievedMessages()` has been exteded to take
   magnetlink messages into account
3. New APIs were added to download torrent data given a magnetlink and
   also to extract messages from downloaded archives, which are then
   later fed to `handleRetrievedMessages`

Closes #2568

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?
- [ ] Have you tested changes with mobile?
- [ ] Have you tested changes with desktop?

0x-r4bbit commented:
This needs tests and some fine-tuning but should be in a reviewable shape.

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -1455,6 +1482,130 @@ func (m *Manager) IsSeedingHistoryArchiveTorrent(communityID types.HexBytes) boo
+	}
+	m.torrentTasks[id] = ml.InfoHash
+
+	<-torrent.GotInfo()
```
So here we're waiting for the magnetlink to emit its metadata (this is needed for us to know what files the torrent will have, so having this info is crucial).

During testing, I've experienced that it sometimes takes a long time to receive a signal, at which point nothing happens here.

cammellos replied:
What happens if you are offline? We should probably timeout this if it could hang forever

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (85)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 81d88dcd | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/1/) | 2022-03-14 12:08:10 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220314-120510-81d88d-pr2585.zip) |
| :heavy_check_mark: | 81d88dcd | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/1/) | 2022-03-14 12:09:40 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220314-120506-81d88d-pr2585.zip) |
| :heavy_check_mark: | 81d88dcd | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/1/) | 2022-03-14 12:11:27 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220314-120506-81d88d-pr2585.aar) |
| | | | | | | |
| :heavy_check_mark: | 7e399636 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/2/) | 2022-03-22 14:18:30 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220322-141510-7e3996-pr2585.zip) |
| :heavy_check_mark: | 7e399636 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/2/) | 2022-03-22 14:19:51 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220322-141505-7e3996-pr2585.aar) |
| :heavy_check_mark: | 7e399636 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/2/) | 2022-03-22 14:20:45 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220322-141505-7e3996-pr2585.zip) |
| | | | | | | |
| :heavy_check_mark: | 83322d23 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/3/) | 2022-03-22 16:28:40 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220322-162706-83322d-pr2585.zip) |
| :heavy_check_mark: | 83322d23 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/3/) | 2022-03-22 16:30:10 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220322-162706-83322d-pr2585.zip) |
| :heavy_check_mark: | 83322d23 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/3/) | 2022-03-22 16:31:40 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220322-162701-83322d-pr2585.aar) |
| | | | | | | |
| :heavy_check_mark: | e8e7cbed | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/4/) | 2022-03-23 18:28:06 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220323-182525-e8e7cb-pr2585.zip) |
| :heavy_check_mark: | e8e7cbed | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/4/) | 2022-03-23 18:30:49 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220323-182527-e8e7cb-pr2585.zip) |
| :heavy_check_mark: | e8e7cbed | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/4/) | 2022-03-23 18:31:10 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220323-182524-e8e7cb-pr2585.aar) |
| | | | | | | |
| :heavy_check_mark: | dfdf2a9c | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/5/) | 2022-03-25 11:30:04 | ~8 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220325-112107-dfdf2a-pr2585.zip) |
| :heavy_check_mark: | dfdf2a9c | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/5/) | 2022-03-25 11:31:11 | ~10 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220325-112102-dfdf2a-pr2585.zip) |
| :heavy_check_mark: | dfdf2a9c | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/5/) | 2022-03-25 11:31:28 | ~10 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220325-112100-dfdf2a-pr2585.aar) |
| | | | | | | |
| :heavy_check_mark: | 1c343520 | [#6](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/6/) | 2022-03-25 17:05:28 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220325-170317-1c3435-pr2585.zip) |
| :heavy_check_mark: | 1c343520 | [#6](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/6/) | 2022-03-26 06:25:22 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220326-062116-1c3435-pr2585.zip) |
| :heavy_check_mark: | 1c343520 | [#6](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/6/) | 2022-03-26 10:56:38 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220326-105317-1c3435-pr2585.aar) |
| | | | | | | |
| :heavy_check_mark: | eb4f9e34 | [#7](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/7/) | 2022-03-28 15:55:08 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220328-155103-eb4f9e-pr2585.zip) |
| :heavy_check_mark: | eb4f9e34 | [#7](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/7/) | 2022-03-28 15:57:18 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220328-155101-eb4f9e-pr2585.aar) |
| :heavy_check_mark: | eb4f9e34 | [#7](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/7/) | 2022-03-28 15:57:30 | ~6 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220328-155105-eb4f9e-pr2585.zip) |
| | | | | | | |
| :heavy_check_mark: | 973d2cad | [#8](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/8/) | 2022-03-28 16:21:45 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220328-161730-973d2c-pr2585.zip) |
| :heavy_check_mark: | 973d2cad | [#8](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/8/) | 2022-03-28 16:24:52 | ~7 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220328-161731-973d2c-pr2585.zip) |
| :heavy_check_mark: | 973d2cad | [#8](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/8/) | 2022-03-28 16:25:13 | ~7 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220328-161730-973d2c-pr2585.aar) |
| | | | | | | |
| :heavy_check_mark: | ca4350c3 | [#9](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/9/) | 2022-04-05 16:53:34 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220405-165155-ca4350-pr2585.zip) |
| :heavy_check_mark: | ca4350c3 | [#9](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/9/) | 2022-04-05 16:55:31 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220405-165153-ca4350-pr2585.aar) |
| :heavy_check_mark: | ca4350c3 | [#9](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/9/) | 2022-04-05 16:56:21 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220405-165155-ca4350-pr2585.zip) |
| | | | | | | |
| :heavy_check_mark: | 7473f5a8 | [#10](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/10/) | 2022-04-06 10:49:07 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220406-104633-7473f5-pr2585.zip) |
| :heavy_check_mark: | 7473f5a8 | [#10](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/10/) | 2022-04-06 10:49:33 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220406-104637-7473f5-pr2585.zip) |
| :heavy_check_mark: | 7473f5a8 | [#10](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/10/) | 2022-04-06 10:50:00 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220406-104633-7473f5-pr2585.aar) |
| :heavy_check_mark: | 7473f5a8 | [#11](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/11/) | 2022-04-06 11:02:17 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220406-110052-7473f5-pr2585.zip) |
| :heavy_check_mark: | 7473f5a8 | [#11](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/11/) | 2022-04-06 11:02:25 | ~1 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220406-110049-7473f5-pr2585.zip) |
| :heavy_check_mark: | 7473f5a8 | [#11](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/11/) | 2022-04-06 11:04:51 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220406-110049-7473f5-pr2585.aar) |
| | | | | | | |
| :heavy_check_mark: | daaf9f6f | [#12](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/12/) | 2022-04-06 11:06:51 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220406-110538-daaf9f-pr2585.zip) |
| :heavy_check_mark: | daaf9f6f | [#12](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/12/) | 2022-04-06 11:08:26 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220406-110537-daaf9f-pr2585.aar) |
| :heavy_check_mark: | daaf9f6f | [#12](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/12/) | 2022-04-06 11:09:59 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220406-110538-daaf9f-pr2585.zip) |
| | | | | | | |
| :heavy_check_mark: | bb7957e6 | [#13](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/13/) | 2022-04-11 10:24:32 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220411-102231-bb7957-pr2585.zip) |
| :heavy_check_mark: | bb7957e6 | [#13](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/13/) | 2022-04-11 10:26:00 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220411-102230-bb7957-pr2585.aar) |
| :heavy_check_mark: | bb7957e6 | [#13](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/13/) | 2022-04-11 10:26:56 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220411-102234-bb7957-pr2585.zip) |
| | | | | | | |
| :heavy_check_mark: | 81b5eab1 | [#14](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/14/) | 2022-04-11 11:47:27 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220411-114140-81b5ea-pr2585.aar) |
| :heavy_check_mark: | 81b5eab1 | [#14](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/14/) | 2022-04-11 11:47:48 | ~6 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220411-114144-81b5ea-pr2585.zip) |
| :heavy_check_mark: | 81b5eab1 | [#14](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/14/) | 2022-04-11 11:53:22 | ~11 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220411-114141-81b5ea-pr2585.zip) |
| | | | | | | |
| :heavy_check_mark: | 57c063d4 | [#15](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/15/) | 2022-04-12 08:48:13 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220412-084606-57c063-pr2585.zip) |
| :heavy_check_mark: | 57c063d4 | [#15](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/15/) | 2022-04-12 08:48:41 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220412-084606-57c063-pr2585.zip) |
| :heavy_check_mark: | 57c063d4 | [#15](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/15/) | 2022-04-12 08:49:30 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220412-084604-57c063-pr2585.aar) |
| | | | | | | |
| :heavy_check_mark: | b2cd3463 | [#16](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/16/) | 2022-04-12 09:22:14 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220412-092035-b2cd34-pr2585.zip) |
| :heavy_check_mark: | b2cd3463 | [#16](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/16/) | 2022-04-12 09:23:59 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220412-092034-b2cd34-pr2585.aar) |
| :heavy_check_mark: | b2cd3463 | [#16](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/16/) | 2022-04-12 09:24:24 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220412-092035-b2cd34-pr2585.zip) |
| | | | | | | |
| :heavy_check_mark: | a6e92f44 | [#17](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/17/) | 2022-04-12 10:12:28 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220412-101005-a6e92f-pr2585.zip) |
| | | | | | | |
| :heavy_check_mark: | c0426c55 | [#18](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/18/) | 2022-04-12 10:20:32 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220412-101540-c0426c-pr2585.zip) |
| :heavy_check_mark: | c0426c55 | [#18](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/18/) | 2022-04-12 10:22:13 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220412-101656-c0426c-pr2585.zip) |
| :heavy_check_mark: | c0426c55 | [#18](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/18/) | 2022-04-12 10:25:16 | ~8 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220412-101657-c0426c-pr2585.aar) |
| | | | | | | |
| :x: | c364972e | [#19](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/19/) | 2022-04-22 07:39:02 | ~1 min | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/19/consoleText) |
| :x: | c364972e | [#19](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/19/) | 2022-04-22 07:39:22 | ~1 min | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/19/consoleText) |
| :x: | c364972e | [#19](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/19/) | 2022-04-22 07:39:57 | ~1 min | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/19/consoleText) |
| | | | | | | |
| :x: | 9d2e6c84 | [#20](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/20/) | 2022-04-22 07:40:44 | ~30 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/20/consoleText) |
| :x: | 9d2e6c84 | [#20](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/20/) | 2022-04-22 07:40:49 | ~35 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/20/consoleText) |
| :x: | 9d2e6c84 | [#20](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/20/) | 2022-04-22 07:40:58 | ~48 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/20/consoleText) |
| | | | | | | |
| :x: | 6ab86430 | [#21](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/21/) | 2022-04-22 07:43:41 | ~28 sec | `ios` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/21/consoleText) |
| :x: | 6ab86430 | [#21](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/21/) | 2022-04-22 07:43:48 | ~35 sec | `linux` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/21/consoleText) |
| :x: | 6ab86430 | [#21](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/21/) | 2022-04-22 07:43:57 | ~48 sec | `android` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/21/consoleText) |
| | | | | | | |
| :heavy_check_mark: | a655450d | [#22](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/22/) | 2022-04-22 07:59:36 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220422-075618-a65545-pr2585.zip) |
| :heavy_check_mark: | a655450d | [#22](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/22/) | 2022-04-22 07:59:46 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220422-075618-a65545-pr2585.zip) |
| :heavy_check_mark: | a655450d | [#22](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/22/) | 2022-04-22 07:59:52 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220422-075615-a65545-pr2585.aar) |
| | | | | | | |
| :heavy_check_mark: | 2682a09e | [#23](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/23/) | 2022-04-22 08:43:37 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220422-084210-2682a0-pr2585.zip) |
| :heavy_check_mark: | 2682a09e | [#23](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/23/) | 2022-04-22 08:44:03 | ~1 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220422-084210-2682a0-pr2585.zip) |
| :heavy_check_mark: | 2682a09e | [#23](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/23/) | 2022-04-22 08:45:19 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220422-084208-2682a0-pr2585.aar) |
| | | | | | | |
| :heavy_check_mark: | 0c8d7986 | [#24](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/24/) | 2022-05-04 12:16:48 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220504-121443-0c8d79-pr2585.zip) |
| :heavy_check_mark: | 0c8d7986 | [#24](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/24/) | 2022-05-04 12:17:40 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220504-121443-0c8d79-pr2585.zip) |
| :heavy_check_mark: | 0c8d7986 | [#24](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/24/) | 2022-05-04 12:18:28 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220504-121442-0c8d79-pr2585.aar) |
| | | | | | | |
| :heavy_check_mark: | 78cdfd08 | [#25](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/25/) | 2022-05-04 13:52:04 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220504-134933-78cdfd-pr2585.zip) |
| :heavy_check_mark: | 78cdfd08 | [#25](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/25/) | 2022-05-04 13:52:44 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220504-134933-78cdfd-pr2585.zip) |
| :heavy_check_mark: | 78cdfd08 | [#25](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/25/) | 2022-05-04 13:53:59 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220504-134931-78cdfd-pr2585.aar) |
| | | | | | | |
| :heavy_check_mark: | 0c201b5b | [#26](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/26/) | 2022-05-06 10:03:14 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220506-095954-0c201b-pr2585.zip) |
| :heavy_check_mark: | 0c201b5b | [#26](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/26/) | 2022-05-06 10:03:42 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220506-095954-0c201b-pr2585.zip) |
| :heavy_check_mark: | 0c201b5b | [#26](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/26/) | 2022-05-06 10:03:55 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220506-095954-0c201b-pr2585.aar) |
| | | | | | | |
| :heavy_check_mark: | 97270282 | [#27](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/27/) | 2022-05-06 10:06:06 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220506-100316-972702-pr2585.zip) |
| :heavy_check_mark: | 97270282 | [#27](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/27/) | 2022-05-06 10:06:42 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220506-100345-972702-pr2585.zip) |
| :heavy_check_mark: | 97270282 | [#27](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/27/) | 2022-05-06 10:06:55 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220506-100358-972702-pr2585.aar) |
| | | | | | | |
| :heavy_check_mark: | ff46fe14 | [#28](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/28/) | 2022-05-09 11:20:21 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220509-111722-ff46fe-pr2585.zip) |
| :heavy_check_mark: | ff46fe14 | [#28](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/28/) | 2022-05-09 11:21:34 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220509-111722-ff46fe-pr2585.zip) |
| :heavy_check_mark: | ff46fe14 | [#28](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/28/) | 2022-05-09 11:23:25 | ~6 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220509-111722-ff46fe-pr2585.aar) |
| | | | | | | |
| :heavy_check_mark: | f7dea6d7 | [#29](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/29/) | 2022-05-09 12:27:22 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220509-122519-f7dea6-pr2585.zip) |
| :heavy_check_mark: | f7dea6d7 | [#29](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/29/) | 2022-05-09 12:27:55 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220509-122519-f7dea6-pr2585.zip) |
| :heavy_check_mark: | f7dea6d7 | [#29](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/29/) | 2022-05-09 12:29:19 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220509-122519-f7dea6-pr2585.aar) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 1b49784c | [#30](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/30/) | 2022-05-10 06:53:03 | ~2 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220510-065009-1b4978-pr2585.zip) |
| :heavy_check_mark: | 1b49784c | [#30](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/30/) | 2022-05-10 06:53:53 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220510-065011-1b4978-pr2585.zip) |
| :heavy_check_mark: | 1b49784c | [#30](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/30/) | 2022-05-10 06:54:15 | ~4 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220510-065008-1b4978-pr2585.aar) |
| | | | | | | |
| :heavy_check_mark: | b3fa1101 | [#31](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-2585/31/) | 2022-05-10 09:13:34 | ~1 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-220510-091139-b3fa11-pr2585.zip) |
| :heavy_check_mark: | b3fa1101 | [#31](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-2585/31/) | 2022-05-10 09:13:36 | ~1 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-220510-091139-b3fa11-pr2585.zip) |
| :heavy_check_mark: | b3fa1101 | [#31](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-2585/31/) | 2022-05-10 09:14:58 | ~3 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-220510-091137-b3fa11-pr2585.aar) |

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -1455,6 +1482,130 @@ func (m *Manager) IsSeedingHistoryArchiveTorrent(communityID types.HexBytes) boo
+	<-torrent.GotInfo()
+	files := torrent.Files()
+
+	indexFile := files[1]
```
We always expect exactly two files `data` and `index`. So accessing `[1]` is safe, but we should probably still dynamically determine, in case we'll add more files to these torrent sin the future (maybe due to https://github.com/status-im/specs/pull/167)

cammellos replied:
Can it be that `files` is of different length? (say 1 or 0), even in cases where the torrent is malformed.
If that's the case, best to error check and return an error, since a panic would just crash the app, and if someone manages to push a malformed torrent for example, it would just crash all clients.

0x-r4bbit replied:
Yea good point. I'll update it so it'll search for an `index` file and if it doesn't exist it'll return early as the torrent is malformed

0x-r4bbit replied:
Updated

0x-r4bbit commented on protocol/messenger_handler.go:
```diff
@@ -626,6 +627,63 @@ func (m *Messenger) HandleCommunityInvitation(state *ReceivedMessageState, signe
+			return m.communitiesManager.UpdateMagnetlinkMessageClock(id, clock)
+		}
+	}
+	return nil
```
Originally I had most of this logic inside `communitiesManaegr`, however because we need access to `handleRetrievedMessages()` I've moved this here. Lemme know if there's a different way to get access to `messenger` APIs inside `communitiesManager` otherwise.

cammellos replied:
You can use pub/sub, but we can have a look at it later

0x-r4bbit commented on protocol/message_persistence.go:
```diff
@@ -1853,7 +1853,7 @@ func (db sqlitePersistence) GetDeletes(messageID string, from string) ([]*Delete
 
 func (db sqlitePersistence) SaveEdit(editMessage EditMessage) error {
-	_, err := db.db.Exec(`INSERT INTO user_messages_edits (clock, chat_id, message_id, text, source, id) VALUES(?,?,?,?,?,?)`, editMessage.Clock, editMessage.ChatId, editMessage.MessageId, editMessage.Text, editMessage.From, editMessage.ID)
+	_, err := db.db.Exec(`INSERT OR REPLACE INTO user_messages_edits (clock, chat_id, message_id, text, source, id) VALUES(?,?,?,?,?,?)`, editMessage.Clock, editMessage.ChatId, editMessage.MessageId, editMessage.Text, editMessage.From, editMessage.ID)
```
This is probably not what we want. I noticed that, when replaying history messages into the DB, message edits were causing SQL errors, saying that the entries have to be unique. So I've added this here to satisfy it.

0x-r4bbit replied:
@cammellos Would be cool if you could confirm whether this here is okay

cammellos replied:
As far as I can see, it should not matter, since all the fields are not modified after insertion, so it would just be replacing the same message. 
Though if you make it not return an error, the code will process the edit multiple times, not sure whether that is an issue (in theory it should be fine), but we need to make sure we test edits before merging, just to make sure.

0x-r4bbit replied:
As per offline discussion, I wanted to changes this to leave out the `OR REPLACE` and check for the error, which can then be interpreted as "don't process this message".

While doing this, I realized I couldn't reproduce the error anymore. So I'll just remove this change for now.

0x-r4bbit commented:
Please review the last commit of this PR

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -87,6 +88,25 @@ func NewManager(identity *ecdsa.PublicKey, db *sql.DB, logger *zap.Logger, verif
+
+func (md archiveMDSlice) Less(i, j int) bool {
+	return md[i].from > md[j].from
+}
```
This is later used when we receive the archive index to make it sortable. This is needed to ensure latest archives are always downloaded first (instead of random).

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -1526,6 +1596,154 @@ func (m *Manager) IsSeedingHistoryArchiveTorrent(communityID types.HexBytes) boo
+		archiveHashes = append(archiveHashes, &archiveMetadata{hash: hash, from: metadata.Metadata.From})
+	}
+
+	sort.Sort(archiveHashes)
```
Here, we're creating the slice of the archive metadata to make it sortable and then sort it, making newer archives come first.

richard-ramos commented on protocol/communities/manager.go:
```diff
@@ -1526,6 +1596,154 @@ func (m *Manager) IsSeedingHistoryArchiveTorrent(communityID types.HexBytes) boo
+func (m *Manager) DownloadHistoryArchivesByMagnetlink(communityID types.HexBytes, magnetlink string) ([]string, error) {
+
+	id := communityID.String()
+	ml, _ := metainfo.ParseMagnetUri(magnetlink)
```
Is it possible that ParseMagnetUri fails if the magnet link somehow is incorrect?
The reason for the question is that it seems the a possible error here is being ignored.

0x-r4bbit replied:
Yeap good point!

richard-ramos commented on protocol/messenger.go:
```diff
@@ -2917,7 +2917,7 @@ func (m *Messenger) RetrieveAll() (*MessengerResponse, error) {
 	}
 
-	return m.handleRetrievedMessages(chatWithMessages)
+	return m.handleRetrievedMessages(chatWithMessages, true)
```
Wouldnt this cause everyone to store the waku messages automatically if they're are admins of a community?
Will this be an option in the future, or should admins be required always to store the messages? 
i.e. I might want to store messages if i'm using my desktop machine, but on mobile i might not.

0x-r4bbit replied:
Right so I had a talk with @John-44 about this and we decided that admins should always store every community message as waku message. However we didn't talk about how this should behave on mobile (where the torrent protocol is turned off by default). So this is a valid concern..

One alternative would be to always put storing the waku messages behind the same flag of whether the protocol is enabled or not. That would apply the same on mobile and desktop.

Just means that, if admins turn off the protocol for whatever reason, they will not store messages as waku messages, resulting in them not being part of the archives.

0x-r4bbit replied:
So to clarify, we're already storing waku messages (also on mobile). This flag here has just been introduced for when we call `handleRetrievedMessages` to replay archive messages into the DB. In that case, we don't want to also store waku messages (in which case the flag is `false`)

richard-ramos commented on protocol/messenger.go:
```diff
@@ -3672,6 +3672,15 @@ func (m *Messenger) handleRetrievedMessages(chatWithMessages map[transport.Filte
+						err = m.HandleHistoryArchiveMagnetlinkMessage(messageState, publicKey, magnetlinkMessage.MagnetUri, magnetlinkMessage.Clock)
+						if err != nil {
+							logger.Warn("failed to handle CommunityMessageArchiveMagnetlink", zap.Error(err))
+							continue
```
```suggestion
							allMessagesProcessed = false
							continue
```

ghost commented:
`vendor/github.com/pion/webrtc/v3/yarn.lock` changed. Pinging @adambabik and @corpetty

0x-r4bbit committed:
Handle history archive magnetlink messages

This introduces the ability for status notes to handle community
history archive magnetlinks. To make this work, a few things are needed:

1. A new database table has been introduced to store message archive
   hashes. This is necessary so status nodes can determine whether or
   not they need to download a certain archive
2. The messenger's `handleRetrievedMessages()` has been exteded to take
   magnetlink messages into account
3. New APIs were added to download torrent data given a magnetlink and
   also to extract messages from downloaded archives, which are then
   later fed to `handleRetrievedMessages`

Closes #2568

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -5,6 +5,7 @@ import (
 	"database/sql"
 	"fmt"
 	"os"
+	"sort"
```
This comment here is just to ensure this file is being reviewed, as github collapses it due to the amount of changes.

0x-r4bbit commented on multiaccounts/accounts/database.go:
```diff
@@ -219,5 +219,5 @@ func (db *Database) AddressExists(address types.Address) (exists bool, err error
 
 func (db *Database) GetNodeConfig() (*params.NodeConfig, error) {
-	return nodecfg.GetNodeConfig(db.db)
+	return nodecfg.GetNodeConfigFromDB(db.db)
```
I needed to make this change because `GeNodeConfig` no longer exists. @richard-ramos is that related to you changes?

richard-ramos replied:
yes, it was surprising that this wasnt caught by the CI

cammellos commented on protocol/communities/manager.go:
```diff
@@ -1530,6 +1600,157 @@ func (m *Manager) IsSeedingHistoryArchiveTorrent(communityID types.HexBytes) boo
+		return nil, err
+	}
+
+	m.logger.Debug("Adding torrent via magnetlink for community", zap.Any("id", id), zap.Any("magnetlink", magnetlink))
```
You can use strings for zap `zap.String("id", id.String())` or `zap.String("magnet", magnetlink.String()), so you get the hexencoded version (otherwise is going to be base encoded)

cammellos commented on protocol/communities/manager.go:
```diff
@@ -1530,6 +1600,157 @@ func (m *Manager) IsSeedingHistoryArchiveTorrent(communityID types.HexBytes) boo
+	m.logger.Debug("Downloading history archive index")
+	indexFile.Download()
+	for {
+		if indexFile.BytesCompleted() == indexFile.Length() {
```
that's a  busy loop, you want to wait you should use a different pattern (use a channel,select/polling etc)

cammellos commented on protocol/communities/manager.go:
```diff
@@ -1530,6 +1600,157 @@ func (m *Manager) IsSeedingHistoryArchiveTorrent(communityID types.HexBytes) boo
+		startIndex := int(metadata.Offset) / pieceLength
+		endIndex := startIndex + int(metadata.Size)/pieceLength
+
+		m.logger.Debug("Downloading data for message archive", zap.Any("hash", hash))
```
we use lowercase for log entries

cammellos commented on protocol/communities/manager.go:
```diff
@@ -1530,6 +1600,157 @@ func (m *Manager) IsSeedingHistoryArchiveTorrent(communityID types.HexBytes) boo
+		}
+
+		for _, message := range archive.Messages {
+			filter := m.transport.FilterByTopic(message.Topic)
```
what are you checking exactly here? meaning, you are checking the topic matches, but generally that's not enough to ensure the message is for you and you should process it. 
Which messages should we process and in which case they should not be processed?

0x-r4bbit replied:
This function (`ExtractMessagesFromHistoryArchive`) aims to create a `map[transport.Filter][]*types.Message` from a downloaded archive, so they can be passed to `handleRetrievedMessages`.

Because we're only getting the topic from each archive message, we need to figure out what filter they belong to, so we can create the correct return type.
All messages extracted from an archive (and passed to `handleRetrievedMessages`) are considered to be processed.

cammellos commented on protocol/messenger_handler.go:
```diff
@@ -683,6 +684,63 @@ func (m *Messenger) HandleCommunityInvitation(state *ReceivedMessageState, signe
+	}
+
+	if m.config.torrentConfig != nil && m.config.torrentConfig.Enabled && settings.HistoryArchiveSupportEnabled {
+		signedByOwnedCommunity, err := m.communitiesManager.IsAdminCommunity(communityPubKey)
```
At a first glance, I think we should be using the key of the community to send this messages, rather than the admins? (there's arguments for both, but worth considering)

0x-r4bbit replied:
So the message with the magnet link **is** signed by the community key.

However this part of the code is handling the magnet link messages. So what we're checking here is: Am I the admin/owner of this community? If yes, then I'm not interested in handling the magnet link because I'm expected to be node that has created the magnet link message from a history archive in the first place.

In other words: only members of the community are interested in this, not admins/owners

cammellos commented on protocol/messenger_handler.go:
```diff
@@ -683,6 +684,63 @@ func (m *Messenger) HandleCommunityInvitation(state *ReceivedMessageState, signe
+		// We are only interested in a community archive magnet link
+		// if it originates from a community that the current account is
+		// part of and doesn't own the private key at the same time
+		if !signedByOwnedCommunity && joinedCommunity && clock >= lastClock {
```
Wondering why we are not interested if we are the admin, is the assumption that we would have the same messages therefore is not useful/is redundant?

0x-r4bbit replied:
As mentioned above, the admin is the one that publishes the history archives via magnetlinks, so that node already has the history. Only members are downloading torrent data

cammellos commented on protocol/messenger_handler.go:
```diff
@@ -683,6 +684,63 @@ func (m *Messenger) HandleCommunityInvitation(state *ReceivedMessageState, signe
+					return
+				}
+
+				_, err = m.handleRetrievedMessages(messagesToHandle, false)
```
This looks a bit dangerous, we currently don't run `handleRetrieveMessages` in parallel, so weird things might happen

0x-r4bbit replied:
Hm.. I wonder what such weird things could be, given that `handleRetrieveMessages` has various checks for incoming messages and handling out of order messages etc.

Any suggestions on how else this should be done?

cammellos replied:
First thing that comes to mind is to have a channel, and handle retrieved messages it's called on it, so you have 2 producers (polling from the network, and torrent) and a single consumer, a bit more complex, but we don't have to think about the code running in parallel, as it might lead to difficult to find bugs.

0x-r4bbit replied:
Okay I'll see if I can make that happen. Thanks for your input!

cammellos replied:
You also need to publish the response, otherwise the client will be in inconsistent state, https://github.com/status-im/status-go/blob/f66ad3d3dd754a8c2e789e3339c5bf0b970129c1/services/ext/service.go#L210 , have you tested this on desktop to see if changes are reflected on the frontend?

0x-r4bbit replied:
In desktop, what we do is we listen for the `HistoryDownloaded` signal, and if that has happened we refetch messages from status-go for the current active channel.

0x-r4bbit committed:
fix: only dispatch magnetlink when owner of community

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -63,8 +65,16 @@ func NewManager(identity *ecdsa.PublicKey, db *sql.DB, logger *zap.Logger, verif
+	archiveLogger, err := lc.Build()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create a archive logger")
+	}
```
I've added a new logger so we get some output in stdout when archives are created/downloaded (easier for debugging, kinda hard when stuff only gets written to geth.log)

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -1530,6 +1609,164 @@ func (m *Manager) IsSeedingHistoryArchiveTorrent(communityID types.HexBytes) boo
+	select {
+	case <-time.After(20 * time.Second):
+		return nil, errors.New("torrent has timed out")
+	case <-torrent.GotInfo():
```
@cammellos I've added a time out of 20 seconds here in case `torrent.GotInfo()` doesn't emit anything

0x-r4bbit commented on protocol/communities/manager.go:
```diff
@@ -1530,6 +1609,164 @@ func (m *Manager) IsSeedingHistoryArchiveTorrent(communityID types.HexBytes) boo
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
```
To make the busy loop less busy, we're now using a 200 ms ticker, which should be more than enough.

0x-r4bbit commented on cmd/statusd/main.go:
```diff
@@ -412,6 +414,13 @@ func retrieveMessagesLoop(messenger *protocol.Messenger, tick time.Duration, can
+			if err != nil {
+				logger.Error("failed to write history archive messages to database", err)
+				continue
+			}
```
@cammellos lemme know if this is what you had in mind

richard-ramos commented on protocol/communities/persistence.go:
```diff
@@ -443,6 +443,11 @@ func (p *Persistence) GetWakuMessagesByFilterTopic(topics []types.TopicType, fro
 }
 
+func (p *Persistence) HasCommunityArchiveInfo(communityID types.HexBytes) (exists bool, err error) {
+	err = p.db.QueryRow(`SELECT count(1) FROM communities_archive_info WHERE community_id = ?`, communityID.String()).Scan(&exists)
```
Not sure if it's worth it, but depending on the number of rows in `communities_archive_info`, `EXISTS` is more efficient
```suggestion
	err = p.db.QueryRow(`SELECT EXISTS (SELECT 1 FROM communities_archive_info WHERE community_id = ?)`, communityID.String()).Scan(&exists)
```

0x-r4bbit replied:
Thanks, this is cool!

richard-ramos commented on protocol/messenger_handler.go:
```diff
@@ -683,6 +684,58 @@ func (m *Messenger) HandleCommunityInvitation(state *ReceivedMessageState, signe
+		return err
+	}
+
+	if m.config.torrentConfig != nil && m.config.torrentConfig.Enabled && settings.HistoryArchiveSupportEnabled {
```
`settings` can be null if communities_settings does not contain the `id`

0x-r4bbit replied:
Good catch

richard-ramos commented on protocol/messenger_handler.go:
```diff
@@ -683,6 +684,58 @@ func (m *Messenger) HandleCommunityInvitation(state *ReceivedMessageState, signe
+		if !signedByOwnedCommunity && joinedCommunity && clock >= lastClock {
+
+			m.communitiesManager.UnseedHistoryArchiveTorrent(id)
+			go func() {
```
Is this a 'slow' operation?  what happens if an user logouts while the goroutine is still executing? If it's important for the goroutine to finish before logout, maybe it's a good idea to add a `sync.WaitGroup` in messenger and wait for it in `func (m *Messenger) Shutdown() (err error)`  (or maybe this is already handled somewhere else?)

0x-r4bbit replied:
This is a very good point. Adding this now

richard-ramos reviewed (approved):
Left some comments, but in general LGTM! :medal_military:

richard-ramos commented:
Please update the VERSION file

0x-r4bbit committed:
Update version number

0x-r4bbit commented:
@cammellos @richard-ramos thanks for all your comments.

- I've updated the SELECT query for checking the existence of archive hash info
- I've added a waitgroup to ensure shutting down the client waits for downloading torrent data to finish
- I've reduced the time interval for checking whether the download of the torrent index file is done to 100ms
- I've introduced a mutex lock for `handleRetrievedMessages` to ensure it doesn't conflict when executed via message archive protocol

churik commented:
If follow the processes (which are still TBD, but better to check twice) we need to check it on the mobile client-side briefly and on the desktop side as well. 
Only in case, you're absolutely sure that it will not introduce any regression issues we can merge it as it is
cc @anastasiyaig

anastasiyaig commented:
@churik u may test it on mobile side and provide feedback. It was not yet tested on desktop, @elina2015 please check this today

0x-r4bbit commented:
> @churik u may test it on mobile side and provide feedback. It was not yet tested on desktop

Just to let you know, this feature is completely turned off on mobile, so the only thing to test/check there is that it indeed doesn't create archives, nor downloads them.

qoqobolo commented:
> Just to let you know, this feature is completely turned off on mobile, so the only thing to test/check there is that it indeed doesn't create archives, nor downloads them.

Hey, thank you for the clarification!

But for context: in such cases, this is not really testing of the feature itself. It was decided (_at least until there’s a better solution, suggestions are still welcome and appriciated_) to create and test clones of Go pull requests in client repositories (Mobile and Desktop) in order to avoid issues that sometimes arise after PRs are merged without any testing on the Go side ([discussion here](https://discord.com/channels/624205794384281629/626738033701683200/971340906001219624)).
So we just run e2e tests, sometimes do some smoke-testing using the corresponding PR in our repo, and then approve the Go PR.

Regarding this particular PR: looks good from the Mobile side, no issues found.
